src/shell/pmi/pmi.c

```diff
@@ -450,28 +433,25 @@ static int init_clique (struct shell_pmi *pmi, const char *opt)
     }
     else {
         shell_log_error ("pmi.clique=%s is invalid", opt);
-        goto error;
+        goto out;
     }
 
     /* Encode to string, and store to local KVS hash.
      */
-
-    /* If value exceeds SIMPLE_KVS_VAL_MAX, skip setting the key
-     * without generating an error.  The client side will not treat
-     * a missing key as an error.  It should be unusual though so log it.
-     */
-    if (pmi_process_mapping_encode (blocks, nblocks, val, sizeof (val)) < 0) {
-        shell_log_errno ("pmi_process_mapping_encode");
+    if (!(s = taskmap_encode_pmi (taskmap))
+        || strlen (s) > SIMPLE_KVS_VAL_MAX) {
+        shell_log_errno ("taskmap_encode_pmi");
         goto out;
     }
```

Comment:
> so I'll test at larger node counts

Oh, duh. The PMI_process_mapping string length for a cyclic distribution doesn't grow that much with the addition of more nodes, just with the addition of more tasks per node.

I did verify that 128 tasks per node and a cyclic distribution causes a fatal error in the PMI plugin. How about if I change it to a non-fatal error something like (instead of requiring the user to re-run with pmi.clique=none)?

```diff
diff --git a/src/shell/pmi/pmi.c b/src/shell/pmi/pmi.c
index 11e4488f1..17a906967 100644
--- a/src/shell/pmi/pmi.c
+++ b/src/shell/pmi/pmi.c
@@ -440,7 +440,10 @@ static int init_clique (struct shell_pmi *pmi, const char *opt)
      */
     if (!(s = taskmap_encode_pmi (taskmap))
         || strlen (s) > SIMPLE_KVS_VAL_MAX) {
-        shell_log_errno ("taskmap_encode_pmi");
+        if (pmi->shell->info->shell_rank == 0)
+            shell_warn ("process mapping exceeds maxlen, "
+                        "using pmi.clique=none");
+        rc = 0;
         goto out;
     }
```

Reply:
Hm, but with `pmi.clique=none` neither MPI or Flux can bootstrap, so I suppose that makes the cyclic case a bit useless?

Reply:
I think since this problem existed before we could tackle it in a separate issue.  A few points
- Flux ought to be able to bootstrap without knowing its taskmap (just use TCP) if it cannot already
- Flux _could_ and probably should bootstrap with the Flux taskmap, if available
- I could be mistaken but I think of the MPIs, only MVAPICH2 _requires_ PMI_process_mapping in order to bootstrap.  I wonder if Slurm can bootstrap a many-task-per-node  cyclic distribution job with MVAPICH2...
- We should consider increasing the PMI max KVS value size to allow cyclic on systems we have today.

Anyway, for a follow-on issue I think.  Cyclic is new functionality for Flux and nothing people are doing now is going to break if we land this as is.

Reply:
> Flux ought to be able to bootstrap without knowing its taskmap (just use TCP) if it cannot already

Oops, I think this was probably broken when the ability for the broker to send its critical ranks upstream was added. I'll open an issue on that.

Reply:
> (I admit I'm a little confused about what should happen if the PMI_process_mapping can't be encoded into SIMPLE_KVS_VAL_MAX)

Sorry I missed this comment.  According to [RFC 13](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_13.html#id14) it should set an empty string in the KVS under that key name.  (Whether that's the right thing or not may be another question, but we should for the time being follow the RFC).

Reply:
The current behavior is to just not set the key in the KVS on overflow:
https://github.com/flux-framework/flux-core/blob/247dc097d26b28fd41aa8a6707ef246405bd3464/src/shell/pmi/pmi.c#L459-L470

The current version of this PR inadvertently turns that into an error (oops, will fix). I guess I'll leave fixing the behavior to comply with the RFC for another PR because I'm worried about unforeseen fallout.